src/modules/cmdpal/extensionsdk/Microsoft.CommandPalette.Extensions.Toolkit/FuzzyStringMatcher.cs

```diff
@@ -0,0 +1,184 @@
+﻿// Copyright (c) Microsoft Corporation
+// The Microsoft Corporation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Globalization;
+
+namespace Microsoft.CommandPalette.Extensions.Toolkit;
+
+// Inspired by the fuzzy.rs from edit.exe
+public static class FuzzyStringMatcher
+{
+    private const int NOMATCH = 0;
+
+    public static int ScoreFuzzy(string needle, string haystack, bool allowNonContiguousMatches = true)
+    {
+        var (s, _) = ScoreFuzzyWithPositions(needle, haystack, allowNonContiguousMatches);
+        return s;
+    }
+
+    public static (int Score, List<int> Positions) ScoreFuzzyWithPositions(string needle, string haystack, bool allowNonContiguousMatches)
+    {
+        if (string.IsNullOrEmpty(haystack) || string.IsNullOrEmpty(needle))
+        {
+            return (NOMATCH, new List<int>());
+        }
+
+        var target = haystack.ToCharArray();
+        var query = needle.ToCharArray();
+
+        if (target.Length < query.Length)
+        {
+            return (NOMATCH, new List<int>());
+        }
+
+        var targetLower = FoldCase(haystack);
+        var queryLower = FoldCase(needle);
+        var targetLowerChars = targetLower.ToCharArray();
+        var queryLowerChars = queryLower.ToCharArray();
```

Comment:
FWIW In my code I use a very cheap arena allocator, but here the GC has to track allocations.

It may further improve performance if you turn this into a class and cache arrays in between calls (since a single needle is presumably used on a lot of haystack strings).

I may also do something similar at some point to avoid recomputing `query` and `queryLower` repeatedly too.